cpp/src/reshape/byte_cast.cu

```diff
@@ -0,0 +1,148 @@
+    uint32_t mask;
+    __device__ void operator()(int byte_index)
+    {
+      d_chars[byte_index] = d_data[byte_index + mask - ((byte_index & mask) << 1)];
+    }
+  };
+
+  template <typename T>
+  std::unique_ptr<column> process(column_view const& input_column,
+                                  flip_endianness configuration,
+                                  rmm::mr::device_memory_resource* mr,
+                                  cudaStream_t stream) const
+  {
+    size_type num_output_elements = input_column.size() * cudf::size_of(input_column.type());
+
+    auto begin          = thrust::make_constant_iterator(cudf::size_of(input_column.type()));
+    auto offsets_column = cudf::strings::detail::make_offsets_child_column(
+      begin, begin + input_column.size(), mr, stream);
+
+    auto byte_column = make_numeric_column(
+      data_type{type_id::UINT8}, num_output_elements, mask_state::UNALLOCATED, stream, mr);
+    auto d_chars = reinterpret_cast<char*>(byte_column->mutable_view().data<T>());
+
+    rmm::device_buffer null_mask = copy_bitmask(input_column, stream, mr);
+
+    if (configuration == flip_endianness::YES) {
+      uint32_t mask = cudf::size_of(input_column.type()) - 1;
+      thrust::for_each(
+        rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator(0),
+        thrust::make_counting_iterator(num_output_elements),
+        FlipEndianness{d_chars, reinterpret_cast<const char*>(input_column.data<T>()), mask});
+    } else {
+      thrust::copy_n(rmm::exec_policy(stream)->on(stream),
+                     reinterpret_cast<const char*>(input_column.data<T>()),
+                     num_output_elements,
+                     d_chars);
+    }
+    return make_lists_column(input_column.size(),
+                             std::move(offsets_column),
+                             std::move(byte_column),
+                             input_column.null_count(),
+                             std::move(null_mask),
+                             stream,
+                             mr);
+  }
+
+  template <typename T,
+            std::enable_if_t<!std::is_integral<T>::value and !is_floating_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& input_column,
+                                     flip_endianness configuration,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream) const
+  {
+    CUDF_FAIL("Unsupported non-numeric and non-string column");
+  }
+
+  template <typename T, std::enable_if_t<is_floating_point<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& input_column,
+                                     flip_endianness configuration,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream) const
+  {
+    auto normalized = normalize_nans_and_zeros(input_column);
```

Comment:
I think we should create a detail version of `normalize_nans_and_zeros` which takes `stream` and `mr` parameters.

Reply:
I'd go even further to say that it's wasteful to materialize the normalized version. The normalization can be done as a transform iterator.

Reply:
It's a little slower to do the endianness flip rather than just recalculating a byte's address, but I guess the cost of the intermediate column far outweighs the performance loss. I'll refactor around doing the normalization and potential flip in the same step.